service/raft/raft_group0.cc

```diff
@@ -488,7 +488,9 @@ static future<bool> synchronize_schema(
         const noncopyable_function<future<bool>()>& can_finish_early,
         abort_source&);
 
-future<> raft_group0::setup_group0(db::system_keyspace& sys_ks, const std::unordered_set<gms::inet_address>& initial_contact_nodes) {
```

Comment:
It seems that this fix is a workaround for those aspects of the current system:
 1) gossip shadow round doesn't update the address map (doesn't fire listeners?)
 2) gossip identifies nodes by IP instead of UUIDs

If the above were fixed, we wouldn't have to add this special-case handling.

We can keep adding such workarounds (I predict there will be more), or fix the gossiper.

Reply:
> gossip shadow round doesn't update the address map (doesn't fire listeners?)

Not firing listeners is by design of the shadow round:
```
                apply_state_locally_without_listener_notification(response.endpoint_state_map).get();
```

The shadow round is designed to be used like this: you call it, retrieve the set of states you need for the startup procedure from the contact point, do whatever you need with these states. It's not supposed to be part of "normal gossiping".

> gossip identifies nodes by IP instead of UUIDs

The bigger problem here is the UI of the replace operation, which is done by passing the IP of the replaced server instead of its ID. We must translate the IP to the Raft ID.

A strategy we could follow to make it less messy is:
1. get https://github.com/scylladb/scylladb/pull/12084 into `master`
2. use Host ID as the Raft ID
3. introduce new UI for the replace operation, where the user must pass the Host ID. Deprecate the old UI. Maybe even disable it completely if Raft is enabled (which could work since Raft will be initially opt-in)

Reply:
It is currently implemented like this (not updating the state), but it doesn't have to. The shadow round is like the real round except that other nodes should not learn about us (we are in the shadow). Why can't we fire listeners here?

We discussed this in the past in https://github.com/scylladb/scylladb/pull/7416#issuecomment-720595534.

Reply:
> The bigger problem here is the UI of the replace operation, which is done by passing the IP of the replaced server instead of its ID. We must translate the IP to the Raft ID.

Maybe it is, but the fact that gossip identifies by IP is also a problem, for this PR, and once nodes are allowed to change IPs.

Reply:
> Why can't we fire listeners here?

No idea, but I guess that someone who implemented the shadow round had some design in mind, they thought about the problem for longer than me and made this decision.

Perhaps we could enable listeners in the shadow round, but we'd need to research the consequences of this change.

I guess I'll just create another issue.

Reply:
https://github.com/scylladb/scylladb/issues/12228

Reply:
> Maybe it is, but the fact that gossip identifies by IP is also a problem, for this PR, and once nodes are allowed to change IPs.

Yes, AFAIK there's a concurrent effort to change this (modify gossiper to identify using an ID). cc @bhalevy

Reply:
We are fading out gossip. Using a dedicated verb to do shadow round (instead of replying on the normal gossip verb handlers) and not fire the listener is one of the early efforts to this goal.  Let's not waste effort in this area.

Reply:
Actually, I think this change is good and in the correct direction because we do not really depend on gossip to get the raft_id (host_id). The verb used by shadow round can be implemented without gossip at all. It is just a verb to get some necessary information from other nodes in the cluster.

Reply:
Going forward, if we can change to make raft_address_map not depending on gossip, we reduce one more dependency on gossip.